src/components/tl/nccl/tl_nccl_coll.c

```diff
@@ -564,6 +565,85 @@ ucc_status_t ucc_tl_nccl_barrier_init(ucc_tl_nccl_task_t *task)
     return UCC_OK;
 }
 
+ucc_status_t ucc_tl_nccl_gather_start(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_nccl_task_t *task   = ucc_derived_of(coll_task, ucc_tl_nccl_task_t);
+    ucc_coll_args_t    *args   = &TASK_ARGS(task);
+    ucc_tl_nccl_team_t *team   = TASK_TEAM(task);
+    ucc_rank_t          rank   = UCC_TL_TEAM_RANK(team);
+    ucc_rank_t          size   = UCC_TL_TEAM_SIZE(team);
+    ucc_ee_h            ee     = coll_task->ee;
+    cudaStream_t        stream = (ee) ? (cudaStream_t) ee->ee_context : team->stream;
+    void               *dst    = args->dst.info.buffer;
+    void               *src    = args->src.info.buffer;
+    ucc_status_t        status = UCC_OK;
+    size_t send_size;
+    ucc_rank_t peer;
+
+    if (rank == args->root) {
+        send_size = ucc_dt_size(args->dst.info.datatype) *
+                    args->dst.info.count / size;
+    } else {
+        send_size = ucc_dt_size(args->src.info.datatype) * args->src.info.count;
+    }
+
+    UCC_TL_NCCL_PROFILE_REQUEST_EVENT(coll_task, "nccl_gather_start", 0);
+    if (rank == args->root) {
+        if (!UCC_IS_INPLACE(*args)) {
+            CUDA_CHECK_GOTO(cudaMemcpyAsync(PTR_OFFSET(dst, rank * send_size),
+                                            src, send_size,
+                                            cudaMemcpyDeviceToDevice, stream),
+                            exit_coll, status);
+        }
+        NCCLCHECK_GOTO(ncclGroupStart(), exit_coll, status,
+                       UCC_TL_TEAM_LIB(team));
+        for (peer = 0; peer < size; peer++) {
+            if (peer == args->root) {
+                continue;
+            }
+            NCCLCHECK_GOTO(ncclRecv(PTR_OFFSET(dst, peer * send_size),
+                                    send_size, ncclChar, peer, team->nccl_comm,
```

Comment:
question - is there a reason to always convert datatype to `ncclChar` as NCCL can support variety of datatypes?